README
NAME
    Module::Install::ForC - the power of M::I for C programs

SYNOPSIS
        # in your Makefile.PL
        use inc::Module::Install;

        my $env = env_for_c(CPPPATH => ['picoev/', 'picohttpparser/']);
        $env->program('testechoclient' => ["testechoclient.c"]);

        WriteMakefile();

        # then, you will get the Makefile:
        all: testechoclient

        clean:
            rm testechoclient testechoclient.o
            rm Makefile

        testechoclient: testechoclient.o
            cc   -fstack-protector -L/usr/local/lib -o testechoclient testechoclient.o

        testechoclient.o: testechoclient.c
            cc -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I picoev/ -I picohttpparser/ -c -o testechoclient.o testechoclient.c

DESCRIPTION
    Module::Install::ForC is a extension library for Module::Install.

    This module provides some useful functions for writing C/C++
    programs/libraries, doesn't depend to Perl.

    M::Install is useful for Perl/XS programming, but this module provides
    M::I power for C/C++ programs!You can use this module as replacement of
    autoconf/automake for easy case.

NOTE
    This is a early BETA release! API will change later.

FUNCTIONS
    is_linux()
    is_mac()
    is_win32()
        Is this the OS or not?

    WriteHeaderForC("config.h")
        Write config.h, contains HAVE_* style definitions generated by
        $env->have_header, $env->have_library.

    my $env = env_for_c(CPPPATH => ['picoev/', 'picohttpparser/']);
        env() returns the instance of M::I::ForC::Env.

        $env contains the build environment variables.The key name is a
        generic value for C.If you want to know about key names, see also
        Module::Install::ForC::Env.

FAQ
    What is supported platform?
        Currently GNU/Linux, OpenSolaris, Mac OSX, and MSWin32.

AUTHOR
    Tokuhiro Matsuno <tokuhirom slkjfd gmail.com>

    mattn(win32 port)

SEE ALSO
    This module is inspired by SCons(<http://www.scons.org/>).

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Module::Install::ForC

0.24

    - support objective-c++
    - removed unused code.
    - move this tests to xt/. because this test fails if c++ compiler is not installed

0.23

    - fixed bug around shared library

0.22

    - fixed packaging issue

0.21

    - hmm fixed M::I related bug

0.20

    - use Module::Install/E::MM directly
    - WriteMakefileForC is deprecated.
    - generate version macro automatically
    - added test case for  Clib.
    - use pm_to_blib
    - added docs for WriteHeaderForC
    - support config.h

0.11

    - support makemaker_args vars
    - support strawberry perl

0.10

    - refactoring and support dmake
    - use EMM->test_via_harness
    - support distclean more.
    - use more thing from E::MM
    - added disttest support

0.09

    - support some vars in Makefile by E::MM
    - added .PHONY
    - support custom postamble

0.08

    - added -g support
    - added $env->enable_warnings
    - these features are stallen from M::I::XSUtil

0.07

    - change test method(INCOMPATIBLE)
    - detect header deps automatically(gcc -MM)
    - check g++ exists using system() on win32.
      (mattn)

0.06

    - updated docs
    - fixed packaging issue
    - fixed g++ detection bug on osx

0.05

    - win32 support(mattn++)
    - do not create a.out file in current dir!

0.04

    - $env->parse_config should return $env itself.

0.03

    - fixed packaging bug.

0.02

    - support solaris

0.01    Thu Oct  1 23:26:40 2009
        - original version


lib/Module/Install/ForC.pm
package Module::Install::ForC;
use strict;
use warnings;
our $VERSION = '0.24';
use 5.008000;
use Module::Install::ForC::Env;
use Config;              # first released with perl 5.00307
use File::Basename ();   # first released with perl 5
use FindBin;             # first released with perl 5.00307

use Module::Install::Base;
our @ISA     = qw(Module::Install::Base);

our %OBJECTS;
our @CONFIG_H;

sub env_for_c {
    my $self = shift;
    $self->_forc_initialize();
    $self->admin->copy_package('Module::Install::ForC::Env');
    Module::Install::ForC::Env->new($self, @_)
}
sub is_linux () { $^O eq 'linux'  }
sub is_mac   () { $^O eq 'darwin' }
sub is_win32 () { $^O eq 'MSWin32' }

# (DEPRECATED)
sub WriteMakefileForC { shift->WriteMakefile(@_) }

sub WriteHeaderForC {
    my ($self, $fname) = @_;
    $fname or die "Usage: WriteHeaderForC('foo_config.h')";

    (my $guard = $fname) =~ tr{a-z./\055}{A-Z___};

    my $header = "#ifndef $guard\n"
               . "#define $guard\n\n";
    my $footer = "\n\n#endif  /* $guard */\n";
    if (my $version = $self->version) {
        (my $verkey = $self->name) =~ s/^Clib-//;
        $verkey =~ tr{a-z./\055}{A-Z___};
        $verkey .= "_VERSION";
        $header .= qq{#define $verkey "$version"\n\n};
    }

    open my $fh, '>', $fname or die "cannot open file($fname): $!";
    print $fh $header . join('', @CONFIG_H) . $footer;
    close $fh;
}

{
    my $initialized = 0;
    sub _forc_initialize {
        return if $initialized++;

        my $self = shift;
        $self->makemaker_args(
            # linking, compiling is job for ForC.
            C      => [],
            OBJECT => '',
        );
    }
}

1;
__END__

=head1 NAME

Module::Install::ForC - the power of M::I for C programs

=head1 SYNOPSIS

    # in your Makefile.PL
    use inc::Module::Install;

    my $env = env_for_c(CPPPATH => ['picoev/', 'picohttpparser/']);
    $env->program('testechoclient' => ["testechoclient.c"]);

    WriteMakefile();

    # then, you will get the Makefile:
    all: testechoclient

    clean:
        rm testechoclient testechoclient.o
        rm Makefile

    testechoclient: testechoclient.o
        cc   -fstack-protector -L/usr/local/lib -o testechoclient testechoclient.o

    testechoclient.o: testechoclient.c
        cc -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I picoev/ -I picohttpparser/ -c -o testechoclient.o testechoclient.c


=head1 DESCRIPTION

Module::Install::ForC is a extension library for Module::Install.

This module provides some useful functions for writing C/C++ programs/libraries, doesn't depend to Perl.

M::Install is useful for Perl/XS programming, but this module provides M::I power for C/C++ programs!You can use this module as replacement of autoconf/automake for easy case.

=head1 NOTE

This is a early BETA release! API will change later.

=head1 FUNCTIONS

=over 4

=item is_linux()

=item is_mac()

=item is_win32()

Is this the OS or not?

=item WriteHeaderForC("config.h")

Write config.h, contains HAVE_* style definitions generated by $env->have_header, $env->have_library.

=item my $env = env_for_c(CPPPATH => ['picoev/', 'picohttpparser/']);

env() returns the instance of M::I::ForC::Env.

$env contains the build environment variables.The key name is a generic value for C.If you want to know about key names, see also L<Module::Install::ForC::Env>.

=back

=head1 FAQ

=over 4

=item What is supported platform?

Currently GNU/Linux, OpenSolaris, Mac OSX, and MSWin32.

=back

=head1 AUTHOR

Tokuhiro Matsuno E<lt>tokuhirom  slkjfd gmail.comE<gt>

mattn(win32 port)

=head1 SEE ALSO

This module is inspired by SCons(L<http://www.scons.org/>).

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
